python_modules/dagster/dagster/_core/definitions/asset_dep.py

```diff
@@ -49,13 +54,19 @@ class AssetDep(
 
     def __new__(
         cls,
-        asset: Union[CoercibleToAssetKey, AssetSpec, AssetsDefinition, SourceAsset],
+        asset: CoercibleToAssetDep,
```

Comment:
From https://github.com/dagster-io/dagster/pull/16432#discussion_r1321989493 

Conclusion is that we should not allow `AssetDep` as a parameter to the `AssetDep` constructor. 

My plan is to then have this constructor be 

```python
def __new__(
        cls,
        asset:  Union[CoercibleToAssetKey, AssetSpec, AssetsDefinition, SourceAsset],
        partition_mapping: Optional[PartitionMapping] = None,
    ):
```